core/src/main/scala/org/apache/spark/util/ManualClock.scala

```diff
@@ -27,6 +27,7 @@ package org.apache.spark.util
 private[spark] class ManualClock(private var time: Long) extends Clock {
 
   private var _isWaiting = false
+  private var _readyForFirstPeek = false
```

Comment:
`_isWaiting` indicates the main `StreamExecution` thread is waiting, while `_readyForFirstPeek` is supposed to indicate whether it's **the first time** the **test thread** knows about the main thread is waiting.

When it's a second time or a third time the test thread knows the main thread is waiting via `isWaitingAndReadyForFirstPeek()`, the test thread itself should block to prevent `advance()`ing too early.